src/fluree/db/query/history.cljc

```diff
@@ -0,0 +1,345 @@
+
+
+(def history-query-validator
+  (m/validator HistoryQuery))
+
+(def history-query-parser
+  (m/parser HistoryQuery))
+
+(defn history-query?
+  "Provide a time range :t and either :history or :commit-details, or both.
+
+  :history - either a subject iri or a vector in the pattern [s p o] with either the
+  s or the p is required. If the o is supplied it must not be nil.
+
+  :context - json-ld context to use in expanding the :history iris.
+
+  :commit-details - if true, each result will have a :commit key with the commit map as a value.
+
+  :t  - a map containing either:
+  - :at
+  - either :from or :to
+
+  accepted values for t maps:
+       - positive t-value
+       - datetime string
+       - :latest keyword "
+  [query]
+  (history-query-validator query))
+
+
+(defn s-flakes->json-ld
+  "Build a subject map out a set of flakes with the same subject.
+
+  {:id :ex/foo :ex/x 1 :ex/y 2}"
+  [db cache compact fuel error-ch s-flakes]
+  (async/go
+    (try*
+      (let [json-chan (json-ld-resp/flakes->res db cache compact fuel 1000000
+                                                {:wildcard? true, :depth 0}
+                                                0 s-flakes)]
+        (-> (<? json-chan)
+            ;; add the id in case the iri flake isn't present in s-flakes
+            (assoc :id (json-ld/compact (<? (dbproto/-iri db (flake/s (first s-flakes)))) compact))))
+      (catch* e
+              (log/error e "Error converting history flakes.")
+              (async/>! error-ch e)))))
+
+(defn t-flakes->json-ld
+  "Build a collection of subject maps out of a set of flakes with the same t.
+
+  [{:id :ex/foo :ex/x 1 :ex/y 2}...]
+  "
+  [db compact cache fuel error-ch t-flakes]
+  (go-try
+    (let [s-flake-partitions (fn [flakes]
+                               (->> flakes
+                                    (group-by flake/s)
+                                    (vals)
+                                    (async/to-chan!)))
+
+          s-ch      (s-flake-partitions t-flakes)
+          s-out-ch  (async/chan)
+          s-json-ch (async/into [] s-out-ch)]
+      (async/pipeline-async 2
+                            s-out-ch
+                            (fn [assert-flakes ch]
+                              (-> (s-flakes->json-ld db cache compact fuel error-ch assert-flakes)
+                                  (async/pipe ch)))
+                            s-ch)
+      (async/<! s-json-ch))))
+
+(defn history-flakes->json-ld
+  "Build a collection of maps for each t that contains the t along with the asserted and
+  retracted subject maps.
+
+  [{:id :ex/foo :f/assert [{},,,} :f/retract [{},,,]]}]
+  "
+  [db context flakes]
+  (go-try
+    (let [fuel    (volatile! 0)
+          cache   (volatile! {})
+
+          compact (json-ld/compact-fn context)
+
+          error-ch   (async/chan)
+          out-ch     (async/chan)
+          results-ch (async/into [] out-ch)
+
+          t-flakes-ch (->> (sort-by flake/t flakes)
```

Comment:
> Also, iirc, these flakes should already be sorted by t, since i'm assuming they came from the tspo index

They won't necessarily always come from tspo. If we're just returning history for a pattern (without commit details), we will use whichever index is appropriate to the input pattern (calculated by `history/history-pattern` [here](https://github.com/fluree/db/blob/feature/commit-metadata/src/fluree/db/query/history.cljc#L192)).

Will the other indexes have out of order `t`s? Now that I think of it I'm not sure that's a problem even for non-`tspo` indexes.